Mathlib/Algebra/Lie/Basic.lean

```diff
@@ -0,0 +1,1134 @@
+  { LieRingModule.compLieHom M f with
+    smul_lie := fun t x m => by
+      simp only [LieRingModule.compLieHom_apply, smul_lie, LieHom.map_smul]
+    lie_smul := fun t x m => by
+      simp only [LieRingModule.compLieHom_apply, lie_smul] }
+#align lie_module.comp_lie_hom LieModule.compLieHom
+
+end ModulePullBack
+
+/-- An equivalence of Lie algebras is a morphism which is also a linear equivalence. We could
+instead define an equivalence to be a morphism which is also a (plain) equivalence. However it is
+more convenient to define via linear equivalence to get `.toLinearEquiv` for free. -/
+structure LieEquiv (R : Type u) (L : Type v) (L' : Type w) [CommRing R] [LieRing L] [LieAlgebra R L]
+  [LieRing L'] [LieAlgebra R L'] extends L →ₗ⁅R⁆ L' where
+  /-- The inverse function of an equivalence of Lie algebras -/
+  invFun : L' → L
+  /-- The inverse function of an equivalence of Lie algebras is a left inverse of the underlying
+  function. -/
+  left_inv : Function.LeftInverse invFun toLieHom.toFun
+  /-- The inverse function of an equivalence of Lie algebras is a right inverse of the underlying
+  function. -/
+  right_inv : Function.RightInverse invFun toLieHom.toFun
+#align lie_equiv LieEquiv
+
+@[inherit_doc]
+notation:50 L " ≃ₗ⁅" R "⁆ " L' => LieEquiv R L L'
+
+namespace LieEquiv
+
+variable {R : Type u} {L₁ : Type v} {L₂ : Type w} {L₃ : Type w₁}
+
+variable [CommRing R] [LieRing L₁] [LieRing L₂] [LieRing L₃]
+
+variable [LieAlgebra R L₁] [LieAlgebra R L₂] [LieAlgebra R L₃]
+
+/-- Consider an equivalence of Lie algebras as a linear equivalence. -/
+def toLinearEquiv (f : L₁ ≃ₗ⁅R⁆ L₂) : L₁ ≃ₗ[R] L₂ :=
+  { f.toLieHom, f with }
+#align lie_equiv.to_linear_equiv LieEquiv.toLinearEquiv
+
+instance hasCoeToLieHom : Coe (L₁ ≃ₗ⁅R⁆ L₂) (L₁ →ₗ⁅R⁆ L₂) :=
+  ⟨toLieHom⟩
+#align lie_equiv.has_coe_to_lie_hom LieEquiv.hasCoeToLieHom
+
+instance hasCoeToLinearEquiv : Coe (L₁ ≃ₗ⁅R⁆ L₂) (L₁ ≃ₗ[R] L₂) :=
+  ⟨toLinearEquiv⟩
+#align lie_equiv.has_coe_to_linear_equiv LieEquiv.hasCoeToLinearEquiv
+
+/-- Coercion to a function. -/
+@[coe] def toFun' (e : L₁ ≃ₗ⁅R⁆ L₂) := e.toFun
+
+/-- see Note [function coercion] -/
+instance : CoeFun (L₁ ≃ₗ⁅R⁆ L₂) fun _ => L₁ → L₂ :=
+  ⟨fun e => e.toFun'⟩
+
+theorem coe_to_lieHom (e : L₁ ≃ₗ⁅R⁆ L₂) : ⇑(e : L₁ →ₗ⁅R⁆ L₂) = e :=
+  rfl
+#align lie_equiv.coe_to_lie_hom LieEquiv.coe_to_lieHom
+
+@[simp]
+theorem coe_to_linearEquiv (e : L₁ ≃ₗ⁅R⁆ L₂) : ⇑(e : L₁ ≃ₗ[R] L₂) = e :=
+  rfl
+#align lie_equiv.coe_to_linear_equiv LieEquiv.coe_to_linearEquiv
+
+@[simp]
+theorem to_linearEquiv_mk (f : L₁ →ₗ⁅R⁆ L₂) (g h₁ h₂) :
+    (mk f g h₁ h₂ : L₁ ≃ₗ[R] L₂) =
+      { f with
+        invFun := g
+        left_inv := h₁
+        right_inv := h₂ } :=
+  rfl
+#align lie_equiv.to_linear_equiv_mk LieEquiv.to_linearEquiv_mk
+
+theorem coe_linearEquiv_injective : Injective ((↑) : (L₁ ≃ₗ⁅R⁆ L₂) → L₁ ≃ₗ[R] L₂) := by
+  rintro ⟨⟨⟨⟨f, -⟩, -⟩, -⟩, f_inv⟩ ⟨⟨⟨⟨g, -⟩, -⟩, -⟩, g_inv⟩
+  intro h
```

Comment:
For reasons I do not understand I cannot use the single invocation:
```lean
rintro ⟨⟨⟨⟨f, -⟩, -⟩, -⟩, f_inv⟩ ⟨⟨⟨⟨g, -⟩, -⟩, -⟩, g_inv⟩ h
```
(same remark applies to the proof of `LieModuleHom.toEquiv_injective`)

Reply:
@digama0 Do you have an idea why this is?